sympy/stats/stochastic_process_types.py

```diff
@@ -0,0 +1,230 @@
+from sympy.core.compatibility import string_types
+from sympy.stats.rv import RandomIndexedSymbol, random_symbols, RandomSymbol
+from sympy.stats.symbolic_probability import Probability
+from sympy.stats.stochastic_process import StochasticPSpace
+
+__all__ = [
+    'StochasticProcess',
+    'DiscreteMarkovChain'
+]
+
+class StochasticProcess(Basic):
+    """
+    Base class for all the stochastic process whether
+    discrete or continuous.
+
+    Parameters
+    ==========
+
+    sym: Symbol or string_types
+    state_space: Set
+        The state space of the stochastic process, by default S.Reals.
+        For discrete sets it is zero indexed.
+    """
+
+    is_Continuous = None
+    is_Discrete = None
+
+    def __new__(cls, sym, state_space=S.Reals):
+        if isinstance(sym, string_types):
+            sym = Symbol(sym)
+        if not isinstance(sym, Symbol):
+            raise TypeError("Name of stochastic process should be either only "
+                            "a string or Symbol.")
+        if isinstance(state_space, set):
+            state_space = list(state_space)
+        if isinstance(state_space, (list, tuple)):
+            state_space = FiniteSet(*state_space)
+        if not isinstance(state_space, Set):
+            raise TypeError("state_space should be an instance of Set/list/tuple/set only.")
+        return Basic.__new__(cls, sym, state_space)
+
+    @property
+    def symbol(self):
+        return self.args[0]
+
+    @property
+    def state_space(self):
+        return self.args[1]
+
+    def __call__(self, time):
+        """
+        Overrided in continuous time stochastic process.
+        """
+        raise NotImplementedError("Use [] for indexing discrete time stochastic process.")
+
+    def __getitem__(self, time):
+        """
+        Overrided in discrete continuous process.
+        """
+        raise NotImplementedError("Use () for indexing continuous time stochastic process.")
+
+    def probability(self, condition):
+        raise NotImplementedError()
+
+class DiscreteMarkovChain(StochasticProcess):
+    """
+    Represents discrete Markov chain.
+
+    Parameters
+    ==========
+
+    sym: Symbol
+    state_space: Set
+        Optional, by default, S.Reals
+    trans_probs: Matrix/ImmutableMatrix/MatrixSymbol
+        Optional, by default, None
+
+    Examples
+    ========
+
+    >>> from sympy.stats.stochastic_process_types import DiscreteMarkovChain
+    >>> from sympy import Matrix, MatrixSymbol, Eq
+    >>> from sympy.stats import P
+    >>> T = Matrix([[0.5, 0.2, 0.3],[0.2, 0.5, 0.3],[0.2, 0.3, 0.5]])
+    >>> Y = DiscreteMarkovChain("Y", [0, 1, 2], T)
+    >>> YS = DiscreteMarkovChain("Y")
+    >>> Y.state_space
+    {0, 1, 2}
+    >>> Y.trans_probs
+    Matrix([
+    [0.5, 0.2, 0.3],
+    [0.2, 0.5, 0.3],
+    [0.2, 0.3, 0.5]])
+    >>> TS = MatrixSymbol('T', 3, 3)
+    >>> P(Eq(YS[3], 2), Eq(YS[1], 1), trans_probs=TS)
```

Comment:
I don't see the point of having this alternate way of providing the transition probabilities. They can be provided as a matrix symbol at the start if that is wanted

Reply:
> Classes with less information can be turned into Markov chains by specifying additional conditions to `P` and `E`.

From https://github.com/sympy/sympy/issues/16895#issuecomment-497649797
@Upabjojr What do you think?

Reply:
I don't think it's wise to add a new parameter to P. The conditions should all be put in the second parameter of P. You can use the & operator to specify the logical and with multiple conditions.

Reply:
I tried but '&' requires all the parameters to be `bool` or `Relational`. Is there any way to pass `Matrix` or `MatrixSymbol` through `&`.

Reply:
@Upabjojr I'm not really sure what you are suggesting. What's the point of TransitionMatrix when the user can just create a new MarkovChain using the matrix?

Reply:
Just putting it as a condition.

Reply:
Okay,  I will make the change today.